Flink: Add the iceberg files committer to collect data files and commit to iceberg table.

openinx opened the pull request:
As we discussed in the pull request https://github.com/apache/iceberg/pull/856,  the flink sink connector actually are composed by two stateful flink operator: 
1.  `FlinkStreamWriter` is used for receiving the records and writing them to `DataFile`(s),  when a checkpoint happen then it will emit the completed `DataFile` to downstream operator. 
2. `FlinkFilesCommitter`: it will collect the `DataFile`  from `FlinkStreamWriter`  and commit the files to iceberg in a transaction, it's a single parallelism operator because we don't expect transaction conflict happen when transaction committing. 

This patch mainly implemented the `FlinkFilesCommitter` and provided several end to end integrate tests to address the correctness.  the dependencies of this one are:  ~~https://github.com/apache/iceberg/pull/1180~~, ~~https://github.com/apache/iceberg/pull/1145~~, ~~https://github.com/apache/iceberg/pull/1175~~

kbendick commented on build.gradle:
```diff
@@ -253,6 +255,9 @@ project(':iceberg-flink') {
     }
+    compileOnly("de.javakaffee:kryo-serializers:0.45") {
+      exclude group: 'com.esotericsoftware', module: 'kryo'
+    }
```
As discussed in https://github.com/apache/iceberg/pull/1145/files#diff-22e0fba47887f8c06bc56a654fd3ee9fL363, we were considering avoiding the usage of `de.javakaffee` for additional kryo serializers.

Could we create custom Flink serializers, or possibly use POJOs or Flink's built in support for Avro serializers. There's a recently posted blog post on the discussion of serializer choices in Flink here: https://flink.apache.org/news/2020/04/15/flink-serialization-tuning-vol-1.html

I think Avro makes the most sense given that it's been native to Flink for a long time and because Avro is so widely used in Iceberg already. 

As an added bonus, Avro supports full schema evolution of state (such that changes to the Avro schema of types stored in stateful operators don't cause said operators to lose their state on upgrade to the new schema, provided they meet certain specifications which are relatively normal rules for type migrations, especially with Avro).

kbendick commented on build.gradle:
```diff
@@ -253,6 +255,9 @@ project(':iceberg-flink') {
+    }
 
     testCompile "org.apache.flink:flink-core"
     testCompile "org.apache.flink:flink-runtime_2.12"
```
Relatively nooby question unrelated to this PR, but does Iceberg only support Flink projects running on Scala 2.12? And is this the same for Spark? For Spark, I would imagine so as Spark 3.0 drops support for Scala 2.11 entirely.

kbendick commented on flink/src/main/java/org/apache/iceberg/flink/IcebergDataStream.java:
```diff
@@ -0,0 +1,109 @@
+
+    operator.addSink(filesCommitter)
+        .name(ICEBERG_FILES_COMMITTER)
+        .uid(UUID.randomUUID().toString())
```
Setting the uid to be a randomly generated number is not a good practice. Operator uid's should be stable, such that between intentional redeployments or simply on restarts due to lost task managers, Flink will know which operators each state belongs to. 

If the uid is changed between restarts, state could be lost.

kbendick replied:
Most of the time, I simply assign the `name` as the `uid`. In Scala, I use an implicit class to assign name and uid at the same time with the same value. This assumes that the name of an operator is unique, but so far I've not run into issues with the handful of jobs that I oversee.

kbendick commented:
Given that https://github.com/apache/iceberg/pull/1180 and https://github.com/apache/iceberg/pull/1175 have been merged, would it be possible for you to merge those commits into this branch to make it easier to review? Or does the dependency on https://github.com/apache/iceberg/pull/1145 make that somewhat burdensome?

Not a major issue, but it would help me in being able to just review / look through the code that's being proposed for addition and not the code that has already been merged in.

If not, no worries. Appreciate all of the work it seems you've put into getting Flink going with Iceberg @openinx. I can see you've submitted quite a number of PRs on the matter so I'm especially looking to learn from you by reading and reviewing these PRs 👍

YesOrNo828 commented on flink/src/main/java/org/apache/iceberg/flink/PartitionKey.java:
```diff
@@ -0,0 +1,126 @@
+    public PartitionKey build(Row row) {
+      Object[] partitionTuple = new Object[size];
+      for (int i = 0; i < partitionTuple.length; i += 1) {
+        partitionTuple[i] = transforms[i].apply(row.getField(pos[i]));
```
`pos[i]` It seemed the fields of flink ddl must be the same sequence with the fields of iceberg schema . what if flink ddl misses some fields? I think partition keys' indices will be different from in iceberg schema

rdblue commented:
@openinx, can you rebase this now that #1145 has been merged?

YesOrNo828 commented on flink/src/main/java/org/apache/iceberg/flink/IcebergFilesCommitter.java:
```diff
@@ -0,0 +1,158 @@
+  @Override
+  public void initializeState(FunctionInitializationContext context) throws Exception {
+    table = TableUtil.findTable(path, conf.get());
+    maxCommittedCheckpointId = parseMaxCommittedCheckpointId(table.currentSnapshot());
```
I think this method `parseMaxCommittedCheckpointId` should be invoked when this job is stored.

openinx replied:
You are right.  here we could only initialize the max-checkpoint-id to -1,  and could not read the checkpoint id from iceberg table if not in a restore job.   Because we may have a table with 3 checkpoints and its max-committed-checkpoint id is 3,  then we stop the current flink job and start another flink job to continue to write this table,  its checkpoint id will start from 1, if we read max-commited-checkpoint id here then we will miss the first three checkpoint's data files in current flink job.

openinx replied:
Added a unit test to address this thing. https://github.com/apache/iceberg/pull/1185/files#diff-67c60f8d1a96e4583f5b53248df15bedR257

YesOrNo828 replied:
Do you have any idea about if the snapshots were removed by another process, and this process commit a new snapshot to iceberg table, the flink committer will not get the correct `maxCommittedCheckpointId `.

openinx replied:
That means someone have expired the latest committed snapshot for a running flink job, usually we flink checkpoint interval will be several minutes, while the snapshot expire interval will be several days or weeks.  It's unlikely that we will expire the latest committed snapshot unless someone set the unreasonable intervals. 

On the other hand,  even if someone removed the latest committed snapshot,  the flink job will write the data to iceberg table correctly unless we restore the flink job once snapshot was removed.

JingsongLi commented:
@openinx and @rdblue I think we can consider https://github.com/apache/iceberg/pull/1332 , what do you think?

JingsongLi commented on flink/src/main/java/org/apache/iceberg/flink/IcebergFilesCommitter.java:
```diff
@@ -0,0 +1,166 @@
+
+  // State for all checkpoints;
+  private static final ListStateDescriptor<byte[]> STATE_DESCRIPTOR =
+      new ListStateDescriptor<>("checkpoints-state", BytePrimitiveArraySerializer.INSTANCE);
```
Can we use Flink `MapTypeInfo`, and addAll to sorted map?

openinx replied:
That sounds good, I've implemented it in this patch:  https://github.com/apache/iceberg/pull/1185/commits/15dd8e0f3f9c294ece18af9686bd469ea3999e68

JingsongLi commented on flink/src/main/java/org/apache/iceberg/flink/IcebergFilesCommitter.java:
```diff
@@ -0,0 +1,166 @@
+
+  @Override
+  public void notifyCheckpointComplete(long checkpointId) {
+    NavigableMap<Long, List<DataFile>> pendingFileMap = dataFilesPerCheckpoint.tailMap(maxCommittedCheckpointId, false);
```
Why exclude `maxCommittedCheckpointId` and `checkpointId`? Just `dataFilesPerCheckpoint.tailMap(checkpointId, true)`? I don't understand why we need store previous checkpoint id.

openinx replied:
I've added few comments about those members here https://github.com/apache/iceberg/pull/1185/commits/6b43ec5b4183d708920e07a4378b8cb6a8ad290b.   

> Why exclude maxCommittedCheckpointId and checkpointId?

The `maxCommittedCheckPointId` 's data files have been committed to iceberg,  so no need to commit them again, otherwise the table will have redundant datas.  For current `checkpointId`,  we've flushed the `dataFilesOfCurrentCheckpoint` into `dataFilesPerCheckpoint` in `snapshotState` method,  so we did not exclude the files for current checkpointId.

JingsongLi replied:
According to the current code, the data of the current checkpoint needs to be committed in the next cycle.
This should not be the ideal way, which can lead to excessive delay.

```
if (maxCommittedCheckpointId >= checkpointId) {
  return;
}
Map<Long, List<DataFile>> pendingFileMap = dataFilesPerCheckpoint.tailMap(checkpointId, true);
List<DataFile> pendingDataFiles = Lists.newArrayList();
for (List<DataFile> dataFiles : pendingFileMap.values()) {
      pendingDataFiles.addAll(dataFiles);
}
-- Do AppendFiles committing
maxCommittedCheckpointId = checkpointId;
pendingFileMap.clear();
```
This is the process I imagined.

openinx replied:
> the data of the current checkpoint needs to be committed in the next cycle.

That's not correct. Since we've merged the `dataFilesOfCurrentCheckpoint` into `dataFilesPerCheckpoint`,  so the  `dataFilesPerCheckpoint.tailMap(maxCommittedCheckpointId, false);`  will contains all the uncommitted data files,  include the files from current checkpoint.

About your described process: 

```java
Map<Long, List<DataFile>> pendingFileMap = dataFilesPerCheckpoint.tailMap(checkpointId, true);
```

This seems won't  commit all the previous files which have committed with failure .  How about the data files between (maxCommittedCheckpointId, checkpointId) ? 

In my thought,  the current process should be correct.  If there're some places I misunderstood, pls correct me,thanks.

JingsongLi replied:
Sorry, wrong reading. Should be`headMap` instead of `tailMap`. You can think again about my above example.

I got your point about `tailMap`. But it is incorrect too. Consider the case:
- 1.snapshot(cpId)
- 2.snapshot(cpId + 1)
- 3.cpId success, notifyComplete(cpId)
- 4.cpId + 1 success, notifyComplete(cpId + 1)

If there is a failure between 3 and 4, the data in table will be incorrect.

openinx replied:
That's a great case (which I did not consider) to proof your point. Yes, we need the `headMap(checkpointId, true)` to avoid future data files being committed to iceberg table,  I've fixed this point in https://github.com/apache/iceberg/pull/1185/commits/598f2353ecad5eda730a1dda96275e552657e466.

openinx commented:
Let me address above comments and fix the failed unit test.

openinx commented:
For this pull request, there are several issue I need to explain here: 
1.  @JingsongLi suggest to consider the two cases: 1.  multiple flink jobs are writing the same table;  2. restart with a new job and continue to write the same time.  For the former case,  we will need a global id, such as job id or application id,  to identify the max checkpoint id we've committed for a given job. For example, we have two flink job: job1 and job2: 
  a.  `job1` commit the iceberg table with maxCheckpointId=1 ; 
  b.  the second `job2` commit the iceberg table with maxCheckpointId=5; 
  c.  `job1` commit the iceberg table with maxCheckpointId=2; 
  d.  the `job2` start to commit again, it need to find the maxCheckpointId corresponding to `job2`, which is 5 now. Then it need to rewrite its maxCheckpointId = 6 and commit the txn. 

The global id of job will also help to resolve the restart issue2, because we will know that the newly started job is starting from checkpoint=1. 

2.  the uid of operator (from @kbendick) issue,  I read the [document](https://ci.apache.org/projects/flink/flink-docs-stable/ops/upgrading.html#matching-operator-state) in flink. Indeed, it's used for state recover and need to be unique across jobs. 

I've address this two issue in the new patch , also attached the unit tests.

JingsongLi commented on flink/src/main/java/org/apache/iceberg/flink/IcebergFilesCommitter.java:
```diff
@@ -0,0 +1,217 @@
+      // checkpoint id should be positive. If it's not positive, that means someone might have removed or expired the
+      // iceberg snapshot, in that case we should throw an exception in case of committing duplicated data files into
+      // the iceberg table.
+      Preconditions.checkArgument(maxCommittedCheckpointId > 0,
```
IIRC, Flink checkpoint id starts at zero. So `maxCommittedCheckpointId >= 0`?

openinx replied:
I checked the [ZooKeeperCheckpointIDCounter](https://github.com/apache/flink/blob/39ef9593f4d88aafaa296f119d9f4a55c9731af1/flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/ZooKeeperCheckpointIDCounter.java#L82) and [StandaloneCheckpointIDCounter](https://github.com/apache/flink/blob/39ef9593f4d88aafaa296f119d9f4a55c9731af1/flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/StandaloneCheckpointIDCounter.java#L33).  Both of them should start from 1.

JingsongLi replied:
Yes, but I think `maxCommittedCheckpointId != INIT_ID` should be more reasonable.

JingsongLi commented on flink/src/main/java/org/apache/iceberg/flink/IcebergFilesCommitter.java:
```diff
@@ -0,0 +1,217 @@
+
+    // Reset the snapshot state to the latest state.
+    checkpointsState.clear();
+    checkpointsState.addAll(ImmutableList.of(dataFilesPerCheckpoint));
```
`checkpointsState.add(dataFilesPerCheckpoint);`

JingsongLi commented on flink/src/main/java/org/apache/iceberg/flink/IcebergFilesCommitter.java:
```diff
@@ -0,0 +1,217 @@
+      pendingDataFiles.forEach(appendFiles::appendFile);
+      appendFiles.set(MAX_COMMITTED_CHECKPOINT_ID, Long.toString(checkpointId));
+      appendFiles.set(GLOBAL_FILES_COMMITTER_UID, filesCommitterUid);
+      appendFiles.commit();
```
Maybe we should commit even `pendingDataFiles` is empty for `MAX_COMMITTED_CHECKPOINT_ID`.

openinx replied:
The `MAX_COMMITTED_CHECKPOINT_ID`  stored in iceberg table is mainly used for avoiding duplicated data files to be committed to iceberg table.  Now we're sure that there's no data files to commit,  advancing the max-committed-checkpoint-id don't have much value in my thought.

JingsongLi replied:
I'm not sure whether the following scenarios are common: however, if there is no data continuously, the previous snapshot may have expired, and an error will be reported after the Flink job failover.
Continue to write snapshots so that others or monitor can feel: Hey, I'm alive, although I don't write data.

openinx replied:
The case you described should be a rare case,  but I agree with you that we'd better to update the max-committed-checkpoint-id, so that the failover won't fail.

JingsongLi commented on flink/src/main/java/org/apache/iceberg/flink/IcebergFilesCommitter.java:
```diff
@@ -0,0 +1,217 @@
+      // checkpoint id should be positive. If it's not positive, that means someone might have removed or expired the
+      // iceberg snapshot, in that case we should throw an exception in case of committing duplicated data files into
+      // the iceberg table.
+      Preconditions.checkArgument(maxCommittedCheckpointId > 0,
```
Maybe should be `checkState`.

JingsongLi commented on flink/src/main/java/org/apache/iceberg/flink/IcebergFilesCommitter.java:
```diff
@@ -0,0 +1,217 @@
+
+  private static final Logger LOG = LoggerFactory.getLogger(IcebergFilesCommitter.class);
+  private static final String GLOBAL_FILES_COMMITTER_UID = "flink.files-committer.uid";
+  private static final String MAX_COMMITTED_CHECKPOINT_ID = "flink.max-committed.checkpoint.id";
```
`checkpoint-id`?

JingsongLi commented on flink/src/main/java/org/apache/iceberg/flink/IcebergFilesCommitter.java:
```diff
@@ -0,0 +1,217 @@
+  private static final long INITIAL_CHECKPOINT_ID = -1L;
+
+  private static final Logger LOG = LoggerFactory.getLogger(IcebergFilesCommitter.class);
+  private static final String GLOBAL_FILES_COMMITTER_UID = "flink.files-committer.uid";
```
uid -> job-id?

JingsongLi commented on flink/src/main/java/org/apache/iceberg/flink/IcebergSinkUtil.java:
```diff
@@ -37,9 +45,32 @@
+        .setParallelism(1)
+        .setMaxParallelism(1);
+
+    return returnStream.addSink(new DiscardingSink()).setParallelism(1);
```
Please name the sink.

JingsongLi commented on flink/src/main/java/org/apache/iceberg/flink/IcebergSinkUtil.java:
```diff
@@ -37,9 +45,32 @@
+                                       String fullTableName,
+                                       Table table,
+                                       TableSchema requestedSchema) {
+    IcebergStreamWriter<RowData> streamWriter = createStreamWriter(table, requestedSchema);
```
I just notice `IcebergStreamWriter` does not set chaining strategy. Should add in the constructor of `IcebergStreamWriter`: `setChainingStrategy(ChainingStrategy.ALWAYS);`

JingsongLi reviewed (changes requested):
Thanks for the update, left some comments.

openinx commented on flink/src/main/java/org/apache/iceberg/flink/RowDataTaskWriterFactory.java:
```diff
@@ -113,7 +114,7 @@ protected PartitionKey partition(RowData row) {
   }
 
-  private static class FlinkFileAppenderFactory implements FileAppenderFactory<RowData> {
+  private static class FlinkFileAppenderFactory implements FileAppenderFactory<RowData>, Serializable {
```
We need to wait until the parquet writer is available. https://github.com/apache/iceberg/pull/1272.  Now only the parquet readers get merged.

openinx commented on core/src/main/java/org/apache/iceberg/io/OutputFileFactory.java:
```diff
@@ -27,7 +28,7 @@
 import org.apache.iceberg.encryption.EncryptionManager;
 
-public class OutputFileFactory {
+public class OutputFileFactory implements Serializable {
```
Because the `IcebergStreamWriter` will need the `RowDataTaskWriterFactory` to create `TaskWriter`,  it could pass it to `TaskWriterFactory` when constructing, that means all members inside `RowDataTaskWriterFactory` should be serializable (include itself), unfortunately `OutputFileFactory` is one of its member.

JingsongLi replied:
`outputFileFactory` should be `transient` in RowDataTaskWriterFactory, it is inited in `initialize`.

openinx replied:
OK, it make sense.

rdblue commented:
I think this one is ready to review and commit, right?

openinx commented:
> I think this one is ready to review and commit, right?

That's right. ( After this DataStream sink,  the next patch is https://github.com/apache/iceberg/pull/1348).

openinx commented:
Rebased the master and did the things btw: 
1.  add ORC as a case in the unit tests; 
2. Since we've introduced the `CatalogLoader` and `TableLoader`, then  we could just use it inside `IcebergFilesCommitter` to load iceberg table lazily,  that looks more reasonable.

JingsongLi commented:
Hi @openinx , for the API level, I think we can do more:
- We can provide two API, one for RowData, one for Row, This can cover not only most users of `DataStream`, but also the requirements of SQL layer.
- Users can just construct a sink from `DataStream` and `TableLoader`, other information can be inferred.

Just like:
```
public class FlinkSink {
  private FlinkSink() {
  }

  public static Builder<Row> forRow(DataStream<Row> input) {
    return new Builder<>(input);
  }

  public static Builder<RowData> forRowData(DataStream<RowData> input) {
    return new Builder<>(input);
  }

  private static class Builder<T> {
    private final DataStream<T> input;
    private TableLoader loader;
    private Configuration hadoopConf;
    private Table table;
    private TableSchema tableSchema;

    // ---------- Required options ------------

    private Builder(DataStream<T> input) {
      this.input = input;
    }

    public Builder tableLoader(TableLoader loader) {
      this.loader = loader;
      return this;
    }

    // ---------- Optional options ------------

    public Builder table(Table newTable) {
      this.table = newTable;
      return this;
    }

    public Builder hadoopConf(Configuration newConf) {
      this.hadoopConf = newConf;
      return this;
    }

    public Builder tableSchema(TableSchema newSchema) {
      this.tableSchema = newSchema;
      return this;
    }

    @SuppressWarnings("unchecked")
    public DataStreamSink<RowData> build() {
      Preconditions.checkNotNull(input, "Input data stream shouldn't be null");
      Preconditions.checkNotNull(loader, "Table loader shouldn't be null");

      if (hadoopConf == null) {
        // load cluster conf
      }

      if (table == null) {
        // load table from table loader
      }

      // tableSchema can be optional

      DataStream<RowData> inputStream;
      Class<T> inputClass = input.getType().getTypeClass();
      if (inputClass == Row.class) {
        DataType type;
        if (tableSchema != null) {
          type = tableSchema.toRowDataType();
        } else {
          type = TypeConversions.fromLogicalToDataType(FlinkSchemaUtil.convert(table.schema()));
        }
        DataStructureConverter converter = DataStructureConverters.getConverter(type);
        inputStream = input.map((MapFunction) converter::toInternal);
      } else if (inputClass == RowData.class) {
        inputStream = (DataStream<RowData>) input;
      } else {
        throw new IllegalArgumentException("Should be Row or RowData");
      }

      // create writer form inputStream.
      // create committer.
      // return DataStreamSink.
      return null;
    }
  }
}
```

JingsongLi commented on flink/src/main/java/org/apache/iceberg/flink/IcebergFilesCommitter.java:
```diff
@@ -0,0 +1,218 @@
+    return new ListStateDescriptor<>("iceberg-files-committer-state", sortedMapTypeInfo);
+  }
+
+  static Long getMaxCommittedCheckpointId(Table table, String flinkJobId) {
```
Actually, the default value of `maxCommittedCheckpointId` can be null instead of `INITIAL_CHECKPOINT_ID`.

openinx replied:
For me, initializing the `maxCommittedCheckpointId`  to  -1 seems simpler to handle, because when calling `SortedMap#tailMap` or `SortedMap#headMap` we don't need to check nullable. The comparasion in `notifyCheckpointComplete` is the similar point.

JingsongLi commented on flink/src/main/java/org/apache/iceberg/flink/IcebergFilesCommitter.java:
```diff
@@ -0,0 +1,218 @@
+  IcebergFilesCommitter(TableLoader tableLoader, Configuration hadoopConf) {
+    this.tableLoader = tableLoader;
+    this.hadoopConf = new SerializableConfiguration(hadoopConf);
+    setChainingStrategy(ChainingStrategy.ALWAYS);
```
We can let committer be a single node.

JingsongLi commented on flink/src/main/java/org/apache/iceberg/flink/IcebergFilesCommitter.java:
```diff
@@ -0,0 +1,218 @@
+    flinkJobId = getContainingTask().getEnvironment().getJobID().toString();
+
+    // Open the table loader and load the table.
+    tableLoader.open(hadoopConf.get());
```
Better to close this `tableLoader` in the `dispose`.

JingsongLi reviewed (commented):
Looks good to me overall, left some minor comments.

openinx commented:
Ping @rdblue , we ( JingsongLi and I ) have reached a basic agreement about this patch.  Would you mind to take a look when you have time ? Thanks.

openinx commented:
Since both parquet reader and writer have been merged, so let me update this patch to add parquet into unit tests.

jrthe42 commented on flink/src/main/java/org/apache/iceberg/flink/FlinkSink.java:
```diff
@@ -0,0 +1,190 @@
+          "Should initialize input DataStream first with DataStream<Row> or DataStream<RowData>");
+      Preconditions.checkArgument(rowInput == null || rowDataInput == null,
+          "Could only initialize input DataStream with either DataStream<Row> or DataStream<RowData>");
+      Preconditions.checkNotNull(table, "Table shouldn't be null");
```
Since we already have TableLoader, will it be easier for user if we load table here ?

openinx replied:
The `table` is used to initialize `IcebergStreamWriter`  at client side,  while the `tableLoader` is used to load table at `TaskManager` side,   using the `tableLoader` for `TaskManager` to load table for  client side looks strange to me.

rdblue commented on core/src/main/java/org/apache/iceberg/io/FileAppenderFactory.java:
```diff
@@ -19,14 +19,15 @@
  * @param <T> data type of the rows to append.
  */
-public interface FileAppenderFactory<T> {
+public interface FileAppenderFactory<T> extends Serializable {
```
This isn't required for Spark. Could we add this to the `FlinkFileAppenderFactory` instead?

openinx replied:
That sounds reasonable.

rdblue commented on flink/src/main/java/org/apache/iceberg/flink/sink/FlinkSink.java:
```diff
@@ -0,0 +1,192 @@
+  }
+
+  public static Builder forRow(DataStream<Row> input) {
+    return new Builder().forRow(input);
```
Curious: are there any plans to support other object models? I know the original sink used Avro, which is supported by Avro and Parquet file formats.

Maybe for that we need a better way to configure file formats with object models, though.

openinx replied:
It will be quite easy to support other object models, we only need a `MapFunction` to convert the generic object to `RowData` and define what's the `TypeInformation` for `RowData`.  For example, we could create a separate builder interfaces for the generic object models on top of this builder: 

```java
  public static <T> Builder builderFor(DataStream<T> input,
                                       Function<T, RowData> converter,
                                       TypeInformation<RowData> outputType) {
    DataStream<RowData> dataStream = input.map(converter::apply, outputType);
    return forRowData(dataStream);
  }
```

rdblue replied:
That sounds okay to me. We can always extend this with a path for other record formats in our version. And I do think it would be much easier to do once we have a more pluggable method to specify object models and data formats.

stevenzwu replied:
@openinx I would prefer to avoid the extra conversion. E.g. we use Avro for streaming data and the `ParquetAvroWriter` from Iceberg library for writer function. On the other hand, we can supply an Avro TaskWriterFactory to `IcebergStreamWriter`.

rdblue commented on flink/src/main/java/org/apache/iceberg/flink/sink/FlinkSink.java:
```diff
@@ -0,0 +1,192 @@
+    @SuppressWarnings("unchecked")
+    public DataStreamSink<RowData> build() {
+      Preconditions.checkArgument(rowInput != null || rowDataInput != null,
+          "Should initialize input DataStream first with DataStream<Row> or DataStream<RowData>");
```
I don't think these messages are very helpful. The problem is that neither `forRow` or `forRowData` is called. Being specific about how to fix it (call one of those methods) would be a more helpful error.

openinx replied:
OK, let me make those more clear.

rdblue commented on flink/src/main/java/org/apache/iceberg/flink/sink/FlinkSink.java:
```diff
@@ -0,0 +1,192 @@
+    }
+
+    private DataStream<RowData> convert() {
+      Preconditions.checkArgument(rowInput != null, "The DataStream<Row> to convert shouldn't be null");
```
The preconditions in build ensure this is never the case. I'm fine keeping the check (up to you) but it seems odd to have it in a private method.

openinx replied:
I think it could be removed now, because we've checked the nullable in the `Builder#build`.

rdblue commented on flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitter.java:
```diff
@@ -0,0 +1,229 @@
+  // <2, <file3>>. Snapshot for checkpoint#1 interrupted because of network/disk failure etc, while we don't expect
+  // any data loss in iceberg table. So we keep the finished files <1, <file0, file1>> in memory and retry to commit
+  // iceberg table when the next checkpoint happen.
+  private final NavigableMap<Long, List<DataFile>> dataFilesPerCheckpoint = Maps.newTreeMap();
```
I think in the Netflix version, we track old checkpoint state by writing a manifest file and appending that manifest to the table. The advantage is that state is really small so the job can be blocked from committing for a long time (days) and can recover when connectivity to commit has been restored.

We use this to handle resilience when AWS regions can't talk to one another and our metastore is in a single region.

JingsongLi replied:
+1 for this approach. There may be thousands of datafiles in a checkpoint for partitioned table.
- In `snapshotState`, `serialize files to Flink state` VS. `serialize files to manifest file`, they should cost about the same.
- In `notifyCheckpointComplete`, just commit a manifest file has better performance instead of commit data files.
- Can reduce the size of States, which are actually put into the memory of the job manager.

Although it is rare that checkpoint fails for a long time, it can improve the robustness of the system.

But, this approach does lead to more code complexity.

openinx replied:
Yeah,  I saw that it will archive all the finished data files into a manifest file and maintain only the manifest files in state backend. (https://github.com/apache/iceberg/pull/856/files#diff-0245903e25b2cace005c52b2c7b130bbR372) .   For my understanding, that's a great point to handle the case that we fail to commit to iceberg table frequently (because we'v buffered all the data files in the manifest files rather than flink state backend),  but I'm no sure whether it is a common case for most of users,  the commit failure should be easily  found if they separated the data files regions from the metastore region,  others also use this way to maitain their data and meta ?  or their states  are enough to maintain those uncommitted data files ?  

If the commit failure is a common case,  I'd happy to contribute to make those data files buffered in a manifest file.   For now,  I think we can make the simple state way go firstly.  Does that make sense ?

rdblue replied:
I completely agree that we should not block adding the sink until this is done. I just wanted to note that this is a good place to put some extra effort. I think that multi-region deployments are common and this can really help. We don't keep Kafka data for a long time, so it is important to us that Flink continues to consume Kafka data to eventually commit to the table, even if it can't commit the manifests right away.

rdblue replied:
@stevenzwu ^^

stevenzwu replied:
commit failure certain isn't common. Agree that it shouldn't be a blocker for the initial version. Just to provide more context on why it is important to us.

Our data warehouses (metastore) only lives in us-east-1 region, while Flink streaming jobs can run in 3 regions (us-east-1, us-west-2, and eu-west-1). 

As Ryan mentioned, we are more concerned about extended outages (like a day) for whatever reason (us-east-1 outage, cross-region network issue, metastore service outage).  This manifest approach allows the Flink jobs to handle those extended outages better.

Flink operator list state can't handle large state well. I vaguely remember 1 or 2 GBs. And it can get pretty slow when the list is large.

openinx replied:
The manifest approach sounds a good improvement to me, it's  a todo issue in the next improvement issue.

rdblue commented on flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitter.java:
```diff
@@ -0,0 +1,229 @@
+    flinkJobId = getContainingTask().getEnvironment().getJobID().toString();
+
+    // Open the table loader and load the table.
+    tableLoader.open(hadoopConf.get());
```
Does this need to be closed?

https://github.com/apache/iceberg/pull/1346/files#r474605890

rdblue replied:
Nevermind, looks like this is done in `dispose()` -- here's a good example of where prefixing with `this.` would be more clear.

rdblue commented on flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitter.java:
```diff
@@ -0,0 +1,229 @@
+    // Open the table loader and load the table.
+    tableLoader.open(hadoopConf.get());
+    table = tableLoader.loadTable();
+    maxCommittedCheckpointId = INITIAL_CHECKPOINT_ID;
```
Minor: I prefer to prefix field names with `this.` when setting instance fields, so it is obvious that it is not a local variable.

rdblue commented on flink/src/main/java/org/apache/iceberg/flink/sink/IcebergStreamWriter.java:
```diff
@@ -34,15 +35,16 @@
   IcebergStreamWriter(String fullTableName, TaskWriterFactory<T> taskWriterFactory) {
     this.fullTableName = fullTableName;
     this.taskWriterFactory = taskWriterFactory;
+    setChainingStrategy(ChainingStrategy.ALWAYS);
```
What does this do?

JingsongLi replied:
The chaining strategy definition:
```
 * Defines the chaining scheme for the operator. When an operator is chained to the
 * predecessor, it means that they run in the same thread. They become one operator
 * consisting of multiple steps.
 *
 * <p>The default value used by the StreamOperator is {@link #HEAD}, which means that
 * the operator is not chained to its predecessor. Most operators override this with
 * {@link #ALWAYS}, meaning they will be chained to predecessors whenever possible.
```

We should also set the chaining strategy of the writer, so as to avoid the network crossing, which will bring additional serialization cost.

openinx replied:
Copied the javadoc from ChainingStrategy here, it's used for chaining the operators to be co-located on the same thread for better performance.
```java
/**
 * Defines the chaining scheme for the operator. When an operator is chained to the
 * predecessor, it means that they run in the same thread. They become one operator
 * consisting of multiple steps.
 *
 * <p>The default value used by the StreamOperator is {@link #HEAD}, which means that
 * the operator is not chained to its predecessor. Most operators override this with
 * {@link #ALWAYS}, meaning they will be chained to predecessors whenever possible.
 */
@PublicEvolving
public enum ChainingStrategy {

	/**
	 * Operators will be eagerly chained whenever possible.
	 *
	 * <p>To optimize performance, it is generally a good practice to allow maximal
	 * chaining and increase operator parallelism.
	 */
	ALWAYS,

	/**
	 * The operator will not be chained to the preceding or succeeding operators.
	 */
	NEVER,

	/**
	 * The operator will not be chained to the predecessor, but successors may chain to this
	 * operator.
	 */
	HEAD
}
```

rdblue replied:
Thanks! Nice to learn more about how Flink works.

rdblue commented:
Overall, I think this is ready to commit. I'd prefer to remove `Serializable` from the appender factory interface first, though. And it would be nice to get some of the other comments fixed or clarified (some are questions since I'm learning about how Flink works). Thanks @openinx!

rdblue commented on flink/src/test/java/org/apache/iceberg/flink/sink/TestFlinkIcebergSink.java:
```diff
@@ -0,0 +1,178 @@
+
+    FlinkSink.forRowData(dataStream)
+        .table(table)
+        .tableLoader(tableLoader)
```
Minor: it seems odd to provide the loader and the table. Couldn't the loader be called in the sink builder?

rdblue commented:
Merged. Thanks @openinx, it is great to have this in. And thanks for helping review, @JingsongLi and @kbendick!

rdblue commented:
What are the next steps for Flink writes? I think we can probably start writing some documentation. The tests for the sink look like this is mostly complete. Maybe docs are an area that @kbendick could help with?

stevenzwu commented on flink/src/main/java/org/apache/iceberg/flink/sink/FlinkSink.java:
```diff
@@ -0,0 +1,220 @@
+      IcebergFilesCommitter filesCommitter = new IcebergFilesCommitter(tableLoader, hadoopConf);
+
+      DataStream<Void> returnStream = rowDataInput
+          .transform(ICEBERG_STREAM_WRITER_NAME, TypeInformation.of(DataFile.class), streamWriter)
```
there can be multiple Iceberg sinks in the same job. we probably should add the table identifier string suffix to make operator name and id unique.

openinx replied:
That sounds good to me,   it's necessary to support multiple iceberg sinks in the same job, I will open an issue and provide a patch with unit tests to address this thing.

stevenzwu commented on flink/src/main/java/org/apache/iceberg/flink/sink/FlinkSink.java:
```diff
@@ -0,0 +1,220 @@
+      DataStream<Void> returnStream = rowDataInput
+          .transform(ICEBERG_STREAM_WRITER_NAME, TypeInformation.of(DataFile.class), streamWriter)
+          .setParallelism(rowDataInput.getParallelism())
+          .transform(ICEBERG_FILES_COMMITTER_NAME, Types.VOID, filesCommitter)
```
committer is a stateful operator, we should probably explicitly set uid.

stevenzwu replied:
never mind. saw earlier comment that this is a todo item.

stevenzwu commented on flink/src/main/java/org/apache/iceberg/flink/sink/FlinkSink.java:
```diff
@@ -0,0 +1,220 @@
+          .setParallelism(1)
+          .setMaxParallelism(1);
+
+      return returnStream.addSink(new DiscardingSink())
```
curious about the reason why don't we make the committer a sink function and instead add a dummy DiscardingSink.

openinx replied:
I think [this](https://github.com/apache/iceberg/pull/1348#discussion_r479859466) would answer your question about why did we use an operator rather than sink function.

stevenzwu replied:
@openinx thx. that does answer my question. 

Still adding a DiscardingSink may confuse the users. It seems that we really need a unified/improved sink interface (similar to FLIP-27) to supported bounded stream in sink function.

What about any other Flink sinks used by bounded streaming/batch job? Do they all have to go through this model?

openinx replied:
There're three cases: 1>  unbounded streaming job; 2> bounded streaming job;  3> batch job.  If users only need the `unbounded streaming` ability, then only need to implement the `SinkFunction`,  otherwise if need both `unbounded streaming` and `bounded streaming`  ability,  then we need to extend/implement the `AbstractStreamOperator` & `BoundedOneInput`  and add the `DiscardingSink` to the tail.  If want to `batch` ability, then  need to provide an `OutputFormat` implementation.  In future flink,  we will unify the case#2 and case#3 in one sink interface,  but for now we have to implement separately for the bounded streaming and batch cases. 

Flink hive connector is a good case, which have support case1, case2 and case3.  It also use the similar way to the current iceberg sink connector now.

stevenzwu replied:
fair enough. we can't solve this problem until Flink improved sink interface.

stevenzwu commented on flink/src/main/java/org/apache/iceberg/flink/sink/FlinkSink.java:
```diff
@@ -0,0 +1,220 @@
+
+      DataStream<Void> returnStream = rowDataInput
+          .transform(ICEBERG_STREAM_WRITER_NAME, TypeInformation.of(DataFile.class), streamWriter)
+          .setParallelism(rowDataInput.getParallelism())
```
this is a good default value for writer parallelism. we have users who want to explicitly control the writer parallelism to control the number of written files. in the future, we may want to allow user to set parallelism in the builder.

openinx replied:
Make sense to me,  could be a following issue.

stevenzwu commented on flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitter.java:
```diff
@@ -0,0 +1,229 @@
+    return new ListStateDescriptor<>("iceberg-files-committer-state", sortedMapTypeInfo);
+  }
+
+  static long getMaxCommittedCheckpointId(Table table, String flinkJobId) {
```
nice. this is the reverse iteration I was looking for. we were just walking through the table.snapshots() iterable.

stevenzwu commented on flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitter.java:
```diff
@@ -0,0 +1,229 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class IcebergFilesCommitter extends AbstractStreamOperator<Void>
```
nit: maybe just "IcebergCommitter". In Iceberg, we can only commit files.

stevenzwu commented on flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitter.java:
```diff
@@ -0,0 +1,229 @@
+  }
+
+  @Override
+  public void notifyCheckpointComplete(long checkpointId) throws Exception {
```
I am wondering if Flink guarantees the serialized execution for notifyCheckpointComplete?

JingsongLi replied:
According to the current implementation, it should be, because task checkpoints are ordered and RPC messages are ordered. But it's better not to rely on this.

stevenzwu replied:
thx. should we raise a jira to track the task?

@JingsongLi you are saying that notifyCheckpointComplete and snapshotState are serialized by the same lock/mutex, right? Otherwise, I can see problem with the concurrent checkpoint handling.

JingsongLi replied:
Yes. You don't need care about thread safety.

stevenzwu commented on flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitter.java:
```diff
@@ -0,0 +1,229 @@
+
+    this.checkpointsState = context.getOperatorStateStore().getListState(STATE_DESCRIPTOR);
+    if (context.isRestored()) {
+      this.maxCommittedCheckpointId = getMaxCommittedCheckpointId(table, flinkJobId);
```
We might have a problem if redeploying the Flink job from external checkpoint. It is a new `flinkJobId` in this case. maxCommittedCheckpointId will be `-1`.  As a result, we can commit those committed files again later.

The way we do de-dup is to generate a hash for the manifest file path and store the hash in snapshot summary. During restore, we use the hash to de-dup if the manifest file was committed or not.

openinx replied:
It's a great point that we could handle, I think we can attach both the `flinkJobId` and `maxCommittedCheckPointId` to the checkpoint state.  when in restoring path,  we read the `flinkJobId` and `maxCommittedCheckpointId` from states, and compare the current flink job id with `flinkJobId` from state.  If the job ids are matched,  then  it is surely be the case that restoring without redploying(case#1), otherwise it's the case you said (case#2).

For case#1, the current code should be correct. 
For case#2, we should use the old `flinkJobId` from state to parse the `maxCommittedCheckpointId` in iceberg table, and use that checkpoint id to filter all the committed data files. 

Does that make sense ?

stevenzwu replied:
I like the proposed solution. it should work.

stevenzwu commented:
nit: on class names

IcebergStreamWriter -> IcebergWriter, because it also BoundedOneInput interface, which seems like a batch job.

IcebergFilesCommitter -> IcebergCommitter. Iceberg only commits files. So including Files seems redundant.

stevenzwu commented on flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitter.java:
```diff
@@ -0,0 +1,229 @@
+
+      SortedMap<Long, List<DataFile>> restoredDataFiles = checkpointsState.get().iterator().next();
+      // Only keep the uncommitted data files in the cache.
+      this.dataFilesPerCheckpoint.putAll(restoredDataFiles.tailMap(maxCommittedCheckpointId + 1));
```
In our implementation, we immediately commit any uncommitted files upon restore to avoid waiting for another checkpoint cycle.

openinx replied:
Committing those uncommitted data files here immediately sounds good to me,  it should won't impact the correctness here.

openinx replied:
Considered this issue again,  if redeploying flink job from external checkpoint, then all data files from `restoredDataFiles`  are from the old flink job, which means its checkpoint id could be range from 1~N.  If we put all those files into `dataFilesPerCheckpoint`, that will introduce the problem:  it did not align with the checkpoint id with the new flink job,  saying when `notifyCheckpointComplete(1)` for the new flink job, it won't  commit all the old data files with checkpointId > 1 to iceberg table.  That's incorrect. 

So it should commit those remaining uncommitted data files to iceberg immediately.

stevenzwu replied:
@openinx if the new job is redeployed from checkpoint N taken by the old job, checkpointId will start from N+1 for the new job. that is my observation of Flink behavior.

openinx replied:
OK , checked the flink savepoint code here,  it was designed as you said. https://github.com/apache/flink/blob/master/flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java#L1371 . 

Thanks for the confirmation,  then I think committing those files in next checkpoint cycle  should also be OK, but there's no harmness for us to commit them immediately.

stevenzwu commented on flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitter.java:
```diff
@@ -0,0 +1,229 @@
+    }
+  }
+
+  private void commitUpToCheckpoint(long checkpointId) {
```
Just want to point out that this will commit if there is zero pending files. For us, we actually still want to commit in this case mainly to update the region watermark info.

openinx replied:
That should be similar to this point from this [comment](https://github.com/apache/iceberg/pull/1185#discussion_r474420170), we actually are doing the logic you said and provided a unit test to address this thing.

stevenzwu commented on flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitter.java:
```diff
@@ -0,0 +1,229 @@
+
+  @Override
+  public void processElement(StreamRecord<DataFile> element) {
+    this.dataFilesOfCurrentCheckpoint.add(element.getValue());
```
this probably gets complicated when we allow concurrent checkpoints. Committer can receive files from both checkpoint N and N+1. We probably need add the checkpointId in the DataFile. I think it might make sense to provide `FlinkDataFile` wrapper so that we can add Flink additional metadata.

We have that in our implementation for low and high timestamps. Now thinking about this issue. Maybe we can include checkpointId too so that committer can distinguish data files from different checkpoints.
```
public class FlinkDataFile implements Serializable {
  private final long lowWatermark;
  private final long highWatermark;
  private final DataFile dataFile;
```

tangchenyang commented:
As a user, how can i use flink write my DataStream to iceberg.
Is there an example or API document or user document for this feature?

openinx commented:
Hi @tangchenyang   The unit test [TestFlinkIcebergSink](https://github.com/apache/iceberg/blob/master/flink/src/test/java/org/apache/iceberg/flink/sink/TestFlinkIcebergSink.java#L133)  provide a complete demo to show you how to write records into iceberg table by flink.  I'm writing a document to show users how to write data into iceberg by both table sql and datastream api, but not finished yet. 
https://github.com/apache/iceberg/pull/1423

tangchenyang commented:
Greate!  Many thanks. @openinx
